JENKINS-41397# validate git credential

vivek opened the pull request:
# Description

See [JENKINS-41397](https://issues.jenkins-ci.org/browse/JENKINS-41397).

Eager validation of credentialId and git uri to fail fast if these values are wrong. I am using 'git ls-remote -h URI> to validate these.

How to test:

### Create Git Pipeline

- Create a  ssh credential thats not supposed to work with GitHub
- Try creating a Git MBP on GitHub with credential created in previous step

curl -X POST -H 'Content-Type: application/json ' -u xx:yyy http://localhost:8080/jenkins/blue/rest/organizations/jenkins/pipelines/

```
{
  "name" : "demo",
  "$class" : "io.jenkins.blueocean.blueocean_git_pipeline.GitPipelineCreateRequest",
  "scmConfig" : {
    "uri" : "git@github.com:vivek/capability-annotation.git",
    "credentialId" : "..."
  }
}
```
you should receive 403 error:

```
{
  "message" : "Failed to create Git pipeline",
  "code" : 403,
  "errors" : [ {
    "message" : "Invalid credentialId: 62f0be4c-bdb2-440b-81de-5255723dd736",
    "code" : "INVALID",
    "field" : "scmConfig.credentialId"
  } ]
}
```
- Try creating a Git MBP on GitHub without credentialId but with bad uri

curl -X POST -H 'Content-Type: application/json ' -u xx:yyy http://localhost:8080/jenkins/blue/rest/organizations/jenkins/pipelines/

```
{
  "name" : "demo",
  "$class" : "io.jenkins.blueocean.blueocean_git_pipeline.GitPipelineCreateRequest",
  "scmConfig" : {
    "uri" : "/sdsd/sdsd/sdsd"
  }
}
```
you should receive 400 error:

```
{
  "message" : "Failed to create Git pipeline",
  "code" : 400,
  "errors" : [ {
    "message" : "Invalid uri: /sdsd/sdsd/sdsd",
    "code" : "INVALID",
    "field" : "scmConfig.uri"
  } ]
}

```

# Submitter checklist
- [x] Link to JIRA ticket in description, if appropriate.
- [x] Change is code complete and matches issue description
- [x] Appropriate unit or acceptance tests or explanation to why this change has no tests
- [x] Reviewer's manual test instructions provided in PR description. See Reviewer's first task below.
- [ ] Ran Acceptance Test Harness against PR changes.

# Reviewer checklist
- [ ] Run the changes and verified the change matches the issue description
- [ ] Reviewed the code
- [ ] Verified that the appropriate tests have been written or valid explanation given

@reviewbybees

vivek commented:
@cliffmeyers PTAL

i386 commented:
NICE 👌

vivek commented:
ATH passes https://ci.blueocean.io/job/ATH-Jenkinsfile/job/master/1149/

michaelneale commented on blueocean-git-pipeline/src/main/java/io/jenkins/blueocean/blueocean_git_pipeline/GitUtils.java:
```diff
@@ -0,0 +1,73 @@
+/**
+ * @author Vivek Pandey
+ */
+class GitUtils {
```
any impact of this on windows ?

vivek replied:
I am using git plugin library calls, that in turn uses JGit so assuming github/git plugins use the same library and they work on windows it should be fine.

michaelneale commented:
Pending cliff - I think this shouldn't be merged until he is happy with it as this changes the behavior of the UI if you put in the wrong thing in a slightly confusing way, until it is changed to look for the response - so provisional :bee: but @cliffmeyers needs to confirm when he wants this in.

cliffmeyers commented:
@vivek I've tested a few different scenarios and receiving slightly different results than I expect.

- There seems to be a difference in behavior between HTTPS and SSH URL's and I am not sure if that's expected. When accessing a Bitbucket private repo via SSH URL, I receive the 403 error as expected (example:  git@bitbucket.org:axltorvalds/axl-jenkins-test-empty-1.git ).  However if I switch to using the HTTPS form of the URL (https://axltorvalds@bitbucket.org/axltorvalds/axl-jenkins-test-empty-1.git ) then I receive the 400 INVALID scmConfig.uri error). Are these supposed to be the same error?
- In both instances above, the pipeline actually gets created and is visible in Jenkins. Is this a bug?
- If I try to create a pipeline as above but the pipeline name is already taken, I receive the 400 error, "ALREADY_EXISTS" for field name. That's expected but it doesn't reveal all the errors at once. Is it possible to combine the "ALREADY_EXISTS" and "INVALID" errors into a single response using the `errors` array? Ideally if the user enters bad URL or picks the wrong credential and has a name conflict, all those errors would be sent back to the UI so we can customize how the errors are presented and let the user fix them all at once without having to first fix the ALREADY_EXISTS error, then fix the INVALID error.

vivek commented:
@cliffmeyers I have addressed all your concerns. PTAL.

Please note that if you access GitHub or bitbucket git  public repo via HTTPS then credential object is ignored. This logic is built underneath git plugin implementation (jgit). I think its bad approach and should file a PR to fix that. till then we have to live with this.

cliffmeyers commented:
@michaelneale per Vivek's comment should we add some helpful text to the Git screen to nudge them away from HTTPS URLs?

cliffmeyers commented:
@vivek just tested this again, still seeing some issues unfortunately:

- When testing against a private Bitbucket SSH URL using a BB username/password credential, I receive a 403 error (INVALID credentialId). I also don't receive an error about a name conflict in this case.
- When testing against a public Github SSH URL with any credential I receive the same error above.
- If I use a bogus URL, I do receive both the INVALID scmConfig.uri error and ALREADY_EXISTS errors - as expected.

In the first two cases above the following error prints to console:
`SEVERE: Error running git remote-ls: Cannot open session, connection is not authenticated.`

Not sure if that is expected or not. Perhaps when you are online we can pair debug this?

vivek commented:
>When testing against a private Bitbucket SSH URL using a BB username/password credential, I receive a 403 error (INVALID credentialId). I also don't receive an error about a name conflict in this case.

Thats expected. With SSH url, it expects ssh credential. 

Reason you don't see name conflict because it would be a 400 and response code can be either 400 or 403:) I am thinking maybe it sends back 400 in all cases, this also fixes underlying git plugin that does not tell 403 vs 400 in all cases of bad credentials?

> When testing against a public Github SSH URL with any credential I receive the same error above.

If you use username/password key with ssh url, 403 is expected. You must use ssh credential in this case.

> In the first two cases above the following error prints to console:

This is intentional, I am logging these errors thrown by git plugin, this is so that user can look at server log to see what error happened.

cliffmeyers commented:
@vivek so just quickly to summarize:

- If a HTTPS URL is used, it must be a public and not require any credentials, as our SCM backend ignores the credential ID.
- If SSH URL is used, the only valid types are SSH private key. This could be a custom private key the user supplied, or the special "SSH key on master" that the user can select.

If I am correct, this is a little unfortunate, as it means that username / password credentials can't be used at all.

I may have asked this before but is it possible for us to decorate the credentials with a capability? If the UI could differentiate between different credential types then we could guide the user to do the right thing based on the type of URL they entered.

vivek commented:
>If a HTTPS URL is used, it must be a public and not require any credentials, as our SCM backend ignores the credential ID.

I think we should continue supplying username/password based credential. We do not know if given repo is public or private.

Git plugin (JGit to be specific) HttpTransport first tries without credential, if it gets 401 then it retries with the supplied credential. So public repos always gets accessed without credentials and due to that can start giving rate limit errors more often. In case of private repo over HTTPS, its unnecessary waste of round trip as private repo will give 401 if no creds are presented.  This needs to be fixed in that plugin and when it gets fixed, supplied credential id will get used. 

> If SSH URL is used, the only valid types are SSH private key. This could be a custom private key the user supplied, or the special "SSH key on master" that the user can select.

Thats right.

> If I am correct, this is a little unfortunate, as it means that username / password credentials can't be used at all.

No credentials will be used if private repos are accessed.

cliffmeyers commented:
Per discussion w/ @vivek, the only change we'll make here is to consolidate all errors under the 400 code. That way the UI can receive multiple errors in a variety of use cases, such as when an invalid credential is supplied (`INVALID` scmConfig.credentialId) *and* the creation would result in a name conflict (`ALREADY_EXISTS`)

vivek commented:
@cliffmeyers All changes from my side done. I assume you are going to make changes on this branch, anyways will wait for your ack to merge to master.

cliffmeyers commented:
@vivek just to confirm, this PR is okay to sit for a while and won't impede progress on any other credential-related PR's? I may take the UI integration work as an opportunity to refine some of the UX. As we discussed, there are some combinations of URL's and credentials that don't make sense, and it would be good to refine the UI to eliminate those.

i386 commented:
Sure thing - whatever you think is best 👍
On Thu, 2 Feb 2017 at 3:14 am, Cliff Meyers <notifications@github.com>
wrote:

> @vivek <https://github.com/vivek> just to confirm, this PR is okay to sit
> for a while and won't impede progress on any other credential-related PR's?
> I may take the UI integration work as an opportunity to refine some of the
> UX. As we discussed, there are some combinations of URL's and credentials
> that don't make sense, and it would be good to refine the UI to eliminate
> those.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/jenkinsci/blueocean-plugin/pull/759#issuecomment-276700745>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AADD7KGaI8jcyqAWfeNSlUS3tAv7rNFXks5rYK9MgaJpZM4LwLUY>
> .
>
-- 
James Dumay
Product Manager
CloudBees, Inc.
Skype: jdumaycb Twitter: i386

vivek commented:
@cliffmeyers 

> As we discussed, there are some combinations of URL's and credentials that don't make sense, and it would be good to refine the UI to eliminate those.

Even without this PR you have same issue no? they would appear aggregated now but they were reported anyways.

Unless it breaks whats already there w.r.t creation flow or otherwise, I would rather have it merged sooner than later as its going to be painful conflicts with credential related changes. Anyways, whatever makes things easy for you, let me know:)

michaelneale commented:
ping @cliffmeyers still want this one?

cliffmeyers commented:
@vivek @michaelneale  I think we should hold off merging this until I make some small changes in the UI first. The UI is only set up to handle the "name conflict" error condition, and since the handling logic is not specific, those errors will trigger the "rename" UI even though that isn't the error. I can get that straightened out tomorrow hopefully so we don't delay this PR too much.

michaelneale commented:
@cliffmeyers ok no problems

cliffmeyers commented:
@vivek @michaelneale I have not been able to pick up this work yet due to the extensive testing on Github creation stuff. Mostly over the hurdle on that now, so I expect to hit this on Monday.

vivek commented:
Latest merge of master passes ATH https://ci.blueocean.io/job/ATH-Jenkinsfile/job/master/1479/. @cliffmeyers  all yours:)

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

kzantow commented on blueocean-git-pipeline/src/test/java/io/jenkins/blueocean/blueocean_git_pipeline/GitScmTest.java:
```diff
@@ -28,7 +39,168 @@ public void setup() throws Exception{
+                        "$class", "io.jenkins.blueocean.blueocean_git_pipeline.GitPipelineUpdateRequest",
+                        "scmConfig", ImmutableMap.of("uri", "/sdsd/sdsd/sdsd")
+                ))
+                .build(Map.class);
```
should this `assert` something?

kzantow replied:
I guess the `RequestBuilder` is doing the validation

vivek replied:
right its asserting on 400 error.

kzantow commented on blueocean-git-pipeline/src/main/java/io/jenkins/blueocean/blueocean_git_pipeline/GitPipelineCreateRequest.java:
```diff
@@ -95,4 +98,45 @@ public BluePipeline create(Reachable parent) throws IOException {
 
+    private void validate(String name, BlueScmConfig scmConfig){
+        if(scmConfig == null){
+            throw new ServiceException.BadRequestExpception(new ErrorMessage(400, "Failed to create Git pipeline")
```
seems like an incorrect error message

vivek replied:
See the line below, this is generic error message. see the line below for field specific error message.

kzantow commented on blueocean-git-pipeline/src/main/java/io/jenkins/blueocean/blueocean_git_pipeline/GitPipelineCreateRequest.java:
```diff
@@ -95,4 +98,45 @@ public BluePipeline create(Reachable parent) throws IOException {
+            errors.add(new ErrorMessage.Error("scmConfig.name", ErrorMessage.Error.ErrorCodes.INVALID.toString(), getName() + "in not a valid name"));
+        }
+
+        if(Jenkins.getInstance().getItem(name)!=null) {
```
is this somehow happening in the context of a folder?

vivek replied:
yes, this is org creation api, so a top level item is checked.

kzantow commented on blueocean-git-pipeline/src/main/java/io/jenkins/blueocean/blueocean_git_pipeline/GitPipelineCreateRequest.java:
```diff
@@ -95,4 +98,45 @@ public BluePipeline create(Reachable parent) throws IOException {
         return null;
     }
 
+    private void validate(String name, BlueScmConfig scmConfig){
```
Is there a way this can be exposed to the REST services so the UI could pre-validate credentials?

vivek replied:
Whats the use case or flow where you need it? with regards to creation flow, it uses credentials API  to  fetch credentials, only valid ones are returned and then its used for rest of the flow.

cliffmeyers commented:
per discussion with @vivek there appear to be some issues with the Git creation API when specifying a credentialId; it returns a "credential not found" error even when the correct ID is specified.

vivek commented:
Looks  like regression resulting from last merge. Looking in to it.

vivek commented:
@cliffmeyers Fixed issue of git pipelines failing to create with valid credential. Also allowed listing of all credentials that user has access to from credentials search API.

i386 commented:
Looks like this is now part of #895. Closing.
